### ISSUE TITLE
RepositoryProvider should expose the dispose callback

### ISSUE THREAD

**Issue by hacker1024** [felangel/bloc#2085]
**Is your feature request related to a problem? Please describe.**
I have a repository that maintains its own HTTP [`Client`](https://pub.dev/documentation/http/latest/http/Client-class.html).
When the repository is no longer needed, this client should be [closed](https://pub.dev/documentation/http/latest/http/Client/close.html). At the moment, this isn't possible.

**Describe the solution you'd like**
I'd like the option to provide a `dispose` callback to the `RepositoryConstructor`.

**Describe alternatives you've considered**
This can be worked around by:
- Using a regular `Provider` (but this needs `package:provider` manually added to the `pubspec.yaml`)
- Using `RepositoryProvider.value` in combination with a `StatefulWidget` (but this is quite inconvenient and slightly more resource-intensive)

**Additional context**

**Comment by narcodico** [comment]
I feel that your repository should never take care of the http client's lifecycle, that's not the purpose nor the responsibility of a repository.
It should rather take in an instance of it and make use of it.
The instance should be managed from outside.

**Comment by ultraon** [comment]
My Team and I are also curious about why RepositoryProvider just suppressed the `dispose` param from Provider.
A use case is straightforward:
We want to create a repository that is scoped on some level and shared between multiple screens (as a Mediator), taking into account that it is not good to have one Bloc dependent on another, and it is not convenient to connect Blocs using events and binging them somewhere outside (in the widget tree), we found that the repository is good enough to share some state (e.g. loaded data, that can be reused on multiple screens). Sometimes repositories have to have streams inside, for example, we subscribed to the Firestore updates inside of the repository, and on each update, we refresh some state inside of the repository. The clients (Blocs) of the repository can listen to the changes and update the representation logic (state), so, the widget tree is updated accordingly until the Bloc is in the widget tree.

Note: the Repository may contain different data sources (network (including WS, SSE), cache, and some platform services (sensors).

It doesn't make sense to have a repository that just wraps API and transforms the raw data into models.

CC: @felangel

**Comment by mrverdant13** [comment]
I believe this feature is worth reconsidering again
There are some cases where repositories manage different kind of inner instances that may take resources that need freed manually.

One of such cases is the use of websockets, where a connection need to be programmatically controlled.

You can check the [VGV's realtime counter example](https://github.com/VGVentures/web_socket_counter_flutter) for the [`web_socket_client` package](https://pub.dev/packages/web_socket_client), where the implementation includes the use of the app widget lifecycle to free the resources of the repository:

```dart
class App extends StatefulWidget {
  const App({super.key});

  @override
  State<App> createState() => _AppState();
}

class _AppState extends State<App> {
  late final CounterRepository _counterRepository;

  @override
  void initState() {
    super.initState();
    _counterRepository = CounterRepository();
  }

  @override
  void dispose() {
    _counterRepository.close();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return RepositoryProvider.value(
      value: _counterRepository,
      child: const AppView(),
    );
  }
}
```

This would be simplified with a `dispose` callback to bind the expected behaviour:

```dart
class App extends StatefulWidget {
  const App({super.key});

  @override
  Widget build(BuildContext context) {
    return RepositoryProvider(
      create: (context) => CounterRepository(),
      dispose: (repository) => repository.close(),
      child: const AppView(),
    );
  }
}
```

**Comment by narcodico** [comment]
Hey @mrverdant13, I believe that's not a great example.

Repositories shouldn't need to be stateful. Specifying a fallback `WebSocket` is creating the need to explicitly dispose of it, which wouldn't happen if the code would require the dependency be passed in from the outside. Then whoever is creating the `WebSocket` is responsible of disposing it, thus avoiding the need of a `close` method in the repository.

Hope that helps!

**Comment by narcodico** [comment]
Hey @ultraon, there are multiple ways of caching your data and none should involve the repository needing to dispose objects. Simply create your dependencies outside and inject them into your repository.
Alternatively you can make use of rxdart operators that are mimicking behavior/replay subjects.

**Comment by ultraon** [comment]
@narcodico I consider `Repository` as a Facade to get data using domain models (business logic models), it may operate by different data sources (in-memory cache, API client, OS-related service, etc.). If we know that the Repository has some scope the same instance should be available on that scope, and whenever the scope is closed all subscriptions (opened sockets) should be disposed inside. With this approach we can share data between different Blocs on different levels, it is one of the recommended ways (by VGV Bloc doc) how to communicate data between different Blocs.
So, the Repository could be disposable and to achieve that we can use a `Provider` from the `provider` package.

**Comment by narcodico** [comment]
@ultraon In my opinion there is a clear difference between a repository and a facade and they shouldn't be used interchangeably nor used for the same purpose.

- the repository pattern is used to manage data access logic and abstract data interactions;
- the facade pattern is used to simplify interaction with multiple complex subsystems by providing a single, unified interface.

So they have different scopes, serve different purpose and deal with different responsibilities.

Nothing stops you from making a repository disposable I just don't think its the way to go. You can obtain your needed architecture very easily while also avoiding a stateful repository.

**Comment by mrverdant13** [comment]
I totally agree with @ultraon 
Reactive caches are another great use case for this feature
Reactive repositories or reactive data sources are a common pattern to share that state among different instances within the app architecture (say data sources, repos or state managers).
Plus, they may also abstract logic that are often atomic and can be self-managed, resulting in a known lifecycle that can benefit from this feature again

Now, let's keep in mind that either the use of the `web_socket_client` package or the `web_socket_channel` package are both implementations where the sockets represent the actual connections, i.e. once they are closed, a new instance is required. 
This implies a method to inject those scoped instances dynamically, which can be cumbersome.

**TL;DR**: Repos may actually hold some state

**Comment by talksik** [comment]
anyone resolve this? My repository does hold state that multiple blocs share

**Comment by ultraon** [comment]
Just use a standard Provider.
